### PR TITLE
Skip unused calculations in sync handler

### PR DESCRIPTION
**Pull request by clokep**
This tweaks some of the calls in the sync handler to avoid calculating things that will not be used.

In particular, if you're not using presence or account lists *and* blocking all rooms you can skip quite a bit of processing. This is useful if you *only* want to-device messages.

Related to #14784, although to really make that better we would need an additional flag for clients to say they're not interested in device list updates. I'll do that as a follow-up.

Based on #14970.

### CODE REVIEW

File: synapse/handlers/sync.py [hunk 1]
```diff
@@ -1832,10 +1839,25 @@ async def _generate_sync_entry_for_rooms(
...
+        ):
+            return set(), set(), set(), set()
+
         # 1. Start by fetching all ephemeral events in rooms we've joined (if required).
         user_id = sync_result_builder.sync_config.user.to_string()
         block_all_room_ephemeral = (
-            since_token is None
+            block_all_rooms
+            or since_token is None
             and sync_result_builder.sync_config.filter_collection.blocks_all_room_ephemeral()
```
Comment (DMRobertson) [inline synapse/handlers/sync.py:1861]: Err, I can never remember the precedence rules. Is this 

1. `(A OR B) AND C` or
2. `A OR (B AND C)`?

(I think 2?)

I'm also not sure what the thinking is behind changing this condition. Is there ever a situation where `block_all_rooms` is True and `sync_result_builder.sync_config.filter_collection.blocks_all_room_ephemeral()` is False?

**Review by DMRobertson — COMMENTED** [review]
I think this looks sane, but I have a few questions and a few comments in passing I'd like to work through with you.

File: synapse/handlers/sync.py [hunk 2]
```diff
@@ -1440,18 +1440,22 @@ async def generate_sync_result(
 
         logger.debug("Fetching room data")
 
+        block_all_presence_data = (
+            since_token is None and sync_config.filter_collection.blocks_all_presence()
+        )
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/sync.py:1445]: > I'm slightly surprised that there's an `and` here rather than an `or`. I wonder if this is correct?

I did not change it in this pull request (as you allude to), it does seem fishy though. I don't understand why we don't return certain things on initial sync TBH.

> Additionally: IIRC there is a Synapse config option to disable presence. I wonder if that should be taken into account here?

Ah, the line below this takes that into account by checking the preference (see below). I think it would be more correct to group that into the single variable though -- otherwise we still risk doing some calculations which will be thrown away.

https://github.com/matrix-org/synapse/blob/f398886ab840061952a4d96ffe2261369c90640a/synapse/handlers/sync.py#L1462-L1471

File: synapse/handlers/sync.py [hunk 3]
```diff
@@ -1908,10 +1915,14 @@ async def handle_room_entries(room_entry: "RoomSyncResultBuilder") -> None:
...
-            newly_left_users,
-        ) = sync_result_builder.calculate_user_changes()
+        if not block_all_presence_data or include_device_list_updates:
+            (
+                newly_joined_or_invited_or_knocked_users,
+                newly_left_users,
+            ) = sync_result_builder.calculate_user_changes()
+        else:
+            newly_joined_or_invited_or_knocked_users = set()
+            newly_left_users = set()
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/sync.py:1942]: A single boolean might be reasonable, I'll check.

For 2:

`newly_joined_or_invited_or_knocked_users` is used in two spots:

* https://github.com/matrix-org/synapse/blob/f398886ab840061952a4d96ffe2261369c90640a/synapse/handlers/sync.py#L1470
* https://github.com/matrix-org/synapse/blob/f398886ab840061952a4d96ffe2261369c90640a/synapse/handlers/sync.py#L1479

And `newly_left_users` in only one:

* https://github.com/matrix-org/synapse/blob/f398886ab840061952a4d96ffe2261369c90640a/synapse/handlers/sync.py#L1481

tl;dr -- I'm confident they only have to do with presence and device list updates.

File: synapse/api/filtering.py [hunk 4]
```diff
@@ -283,6 +283,9 @@ async def filter_room_account_data(
             await self._room_filter.filter(events)
         )
 
+    def blocks_all_rooms(self) -> bool:
+        return self._room_filter.filters_all_rooms()
+
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/api/filtering.py:288]: This unfortunately matches the format of all of the filtering code FTR. We should probably make that code more robust, but don't want to handle that here.

File: synapse/handlers/sync.py [hunk 5]
```diff
@@ -1837,10 +1839,25 @@ async def _generate_sync_entry_for_rooms(
...
+
+        # 0. If there are no rooms to return *and* we don't care about presence
+        # or device list updates, there's nothing to do.
+        if (
+            block_all_rooms
+            and block_all_presence_data
+            and not include_device_list_updates
+        ):
+            return set(), set(), set(), set()
+
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/sync.py:1855]: Maybe? The 2nd clause though would still come into effect for the device lists / presence stuff (i.e. the "if nothing has changed there's nothing to do"). Maybe by `joined` being empty though this would work fine without any other changes.

I'm hesitant to change that here!

File: synapse/handlers/sync.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/sync.py:1445]: I did the refactor for the config flag and also inverted the logic so it is "include" instead of "block all" (which makes my head hurt less) as a separate PR: #14970.

File: synapse/handlers/sync.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/handlers/sync.py:1855]: Very fair.

I hate the sync logic.

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/sync.py:1855]: OK, so I started combining the booleans as you asked in https://github.com/matrix-org/synapse/pull/14908#discussion_r1088216958; then realized it should just be hoisted to the caller.

File: synapse/handlers/sync.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/sync.py]: This code got simplified so, doesn't matter.

**Review by DMRobertson — APPROVED** [review]
Many thanks for this. The sync logic is not pretty but I think this is good incremental progress.
